File: frontend/lib/types/document.ts
/** Home Digital Twin types — mirror backend/app/schemas/document.py + service_record.py
 * and the `docs` / `service_records` tables. One flat shape per concept (not a
 * class per category) since the DB stores category-specific fields in a single
 * metadata JSONB column rather than a table per category.
 */

export type DocumentCategory = "invoice" | "warranty" | "photo" | "manual";

export interface HomeDocument {
  id: string;
  category: DocumentCategory;
  title: string;
  fileUrl: string | null;
  fileType: string;
  uploadedAt: string;
  tags: string[];
  linkedAppliance: string | null;
  notes: string | null;
  // Invoice-specific (also usable elsewhere where relevant)
  vendor?: string | null;
  amount?: number | null;
  purchaseDate?: string | null;
  orderNumber?: string | null;
  // Warranty-specific
  brand?: string | null;
  expiresAt?: string | null;
}

/** Fields the "Add document" form can submit. No real file upload yet — see
 * Phase 1 of the Digital Twin plan (object storage isn't wired up). */
export interface HomeDocumentCreate {
  category: DocumentCategory;
  title: string;
  fileType?: string;
  tags?: string[];
  linkedAppliance?: string | null;
  notes?: string | null;
  vendor?: string | null;
  amount?: number | null;
  purchaseDate?: string | null;
  orderNumber?: string | null;
  brand?: string | null;
  expiresAt?: string | null;
}

export interface ServiceRecord {
  id: string;
  serviceDate: string;
  contractorName: string | null;
  workPerformed: string;
  cost: number | null;
  linkedDocumentId: string | null;
  notes: string | null;
}

export interface ServiceRecordCreate {
  serviceDate: string;
  contractorName?: string | null;
  workPerformed: string;
  cost?: number | null;
  linkedDocumentId?: string | null;
  notes?: string | null;
}

File: backend/app/homeowner/schemas/document.py
"""Home Digital Twin document schemas — mirror frontend/lib/homeowner/types/document.ts.

Backed by a single `docs` table (see backend/db/schema.sql) — one shape for
Invoices/Warranty Cards/Photos/Technical Manuals, with category-specific
fields kept optional rather than one table per category.
"""

from typing import Literal

from app.shared.schemas.user import CamelModel

DocumentCategory = Literal["invoice", "warranty", "photo", "manual"]


class DocumentOut(CamelModel):
    id: str
    category: DocumentCategory
    title: str
    file_url: str | None = None
    file_type: str
    uploaded_at: str
    tags: list[str] = []
    linked_appliance: str | None = None
    notes: str | None = None
    # Invoice
    vendor: str | None = None
    amount: int | None = None
    purchase_date: str | None = None
    order_number: str | None = None
    # Warranty / manual
    brand: str | None = None
    expires_at: str | None = None


class DocumentCreate(CamelModel):
    category: DocumentCategory
    title: str
    file_type: str = "pdf"
    tags: list[str] = []
    linked_appliance: str | None = None
    notes: str | None = None
    vendor: str | None = None
    amount: int | None = None
    purchase_date: str | None = None
    order_number: str | None = None
    brand: str | None = None
    expires_at: str | None = None
